### ISSUE TITLE
Dataverse Page - 4.6.2 Regression #1 - Dataset Cards/Thumbnails

### ISSUE THREAD

**Issue by raprasad** [IQSS/dataverse#3852]
Scenario:
- Local dataverse system running develop as of 5/24/2017
- Root dataverse page.  System has:
  - 1 Dataverse (root)
  - 4 Datasets (these are heavily versioned)
  - 13 files 
### Each home page refresh runs 890 queries 
  - compared to **78 queries in 4.6.1**

## These 3 queries are run 200 times each

```sql
(1) COUNT: 200   QUERY(truncated): SELECT ID, UNF, ARCHIVENOTE, ARCHIVETIME, 
CREATETIME, DEACCESSIONLINK, INREVIEW, LASTUPDATETIME, MINORVERSIONNUMBER,
 RELEASETIME, VERSION, VERSIONNOTE, VERSIONNUMBER, VERSIONSTATE, DATASET_ID, termsOfUseAndAccess_id 
FROM DATASETVERSION WHERE (termsOfUseAndAccess_id = ?)
----------------------------------------
(2) COUNT: 200   QUERY(truncated): SELECT ID, CREATETIME, NAME, USAGECOUNT, 
DATAVERSE_ID, termsOfUseAndAccess_id
 FROM TEMPLATE WHERE (termsOfUseAndAccess_id = ?)
----------------------------------------
(3) COUNT: 200   QUERY(truncated): SELECT ID, AVAILABILITYSTATUS, CITATIONREQUIREMENTS,
 CONDITIONS, CONFIDENTIALITYDECLARATION, CONTACTFORACCESS, DATAACCESSPLACE, 
DEPOSITORREQUIREMENTS, DISCLAIMER, FILEACCESSREQUEST, LICENSE, ORIGINALARCHIVE, 
RESTRICTIONS, SIZEOFCOLLECTION, SPECIALPERMISSIONS, STUDYCOMPLETION, 
TERMSOFACCESS, TERMSOFUSE FROM TERMSOFUSEANDACCESS WHERE (ID = ?)
```

**Comment by raprasad** [comment]
@kcondon @pdurbin 

### This appears to have been introduced *after* 4.6.1

Scenario:
- Same data as described above
- Downloaded, compiled ran 4.6.1 against the same database 


### Each home page refresh runs between 78 and 101 queries*
  - (* logged and not-logged in, respectively)

```sql
----------------------------------------
(1) COUNT: 20   QUERY(truncated): SELECT t0.ID, t0.DESCRIPTION, t0.DISPLAYNAME, t0.GROUPALIAS, t0.GROUPALIASINOWNER, t0.OWNER_ID FROM EXPLICITGROUP t0, ExplicitGroup_CONTAINEDROLEASSIGNEES t1 WHERE (((t0.OWNER_ID = ?) AND (t1.CONTAINEDROLEASSIGNEES = ?)) AND (t1.ExplicitGroup_ID = t0.ID))
----------------------------------------
(2) COUNT: 18   QUERY(truncated): SELECT NAME, CONTENT FROM SETTING WHERE (NAME = ?)
----------------------------------------
(3) COUNT: 13   QUERY(truncated): SELECT ID, DISPLAYNAME, NAME, owner_id FROM METADATABLOCK WHERE (ID = ?)
----------------------------------------
(4) COUNT: 10   QUERY(truncated): SELECT ID, ASSIGNEEIDENTIFIER, PRIVATEURLTOKEN, DEFINITIONPOINT_ID, ROLE_ID FROM ROLEASSIGNMENT WHERE ((ASSIGNEEIDENTIFIER = ?) AND (DEFINITIONPOINT_ID = ?))
----------------------------------------
(5) COUNT: 6   QUERY(truncated): SELECT ID, ADVANCEDSEARCHFIELDTYPE, ALLOWCONTROLLEDVOCABULARY, ALLOWMULTIPLES, description, DISPLAYFORMAT, DISPLAYONCREATE, DISPLAYORDER, FACETABLE, FIELDTYPE, name, REQUIRED, title, WATERMARK, METADATABLOCK_ID, PARENTDATASETFIELDTYPE_ID FROM DATASETFIELDTYPE WHERE (ID = ?)
----------------------------------------
(6) COUNT: 4   QUERY(truncated): SELECT t0.ID, t0.DESCRIPTION, t0.DISPLAYNAME, t0.GROUPALIAS, t0.GROUPALIASINOWNER, t0.OWNER_ID FROM EXPLICITGROUP t0, ExplicitGroup_CONTAINEDROLEASSIGNEES t1 WHERE ((t1.CONTAINEDROLEASSIGNEES = ?) AND (t1.ExplicitGroup_ID = t0.ID))
----------------------------------------
(7) COUNT: 2   QUERY(truncated): SELECT ID, ADVANCEDSEARCHFIELDTYPE, ALLOWCONTROLLEDVOCABULARY, ALLOWMULTIPLES, description, DISPLAYFORMAT, DISPLAYONCREATE, DISPLAYORDER, FACETABLE, FIELDTYPE, name, REQUIRED, title, WATERMARK, METADATABLOCK_ID, PARENTDATASETFIELDTYPE_ID FROM DATASETFIELDTYPE WHERE (name = ?)
```

**Comment by pdurbin** [comment]
@raprasad bleh! Thanks for checking if it's a regression or not.

**Comment by raprasad** [comment]
For the scenario in initial description, the cause is these 5 lines in the SearchServiceBean:

https://github.com/IQSS/dataverse/commit/9ddd0a4f71e2dbe3eef4cc28f40395d4b6053936#diff-b55512aefa9f8bcc2670df32e065f3dcR465

```java
                DvObject dvObject = dvObjectService.findDvObject(entityid);
                if (dvObject != null) {
                    Dataset dataset = (Dataset) dvObject;
                    solrSearchResult.setDatasetThumbnail(dataset.getDatasetThumbnail());
                }
```

Those 5 lines eventually call the ```DatasetUtil``` file and the method ```getThumbnail(Dataset dataset)```

---

The test page currently creates 922 total queries--the majority of which are:

- 39 queries: ```DvObject dvObject = dvObjectService.findDvObject(entityid);```
- 781 queries: ```solrSearchResult.setDatasetThumbnail(dataset.getDatasetThumbnail());```
   - All queries are from: ```dataset.getDatasetThumbnail()``` in the line above
   - next call: ```Dataset``` : ```getDatasetThumbnail()```

## problem area -- or maybe it shouldn't be called at all for Dataverse page rendering
   - eventual culprit: -> ```DatasetUtil``` : ```getThumbnail(Dataset dataset)```

**Comment by raprasad** [comment]
### More info on ```DatasetUtil : getThumbnail(Dataset dataset)```

My latest test page creates **922** queries.

This line is responsible for **437** of those queries:

```java
String title = dataset.getLatestVersion().getTitle();
```

#### Breakdown:
  -  ```.getLatestVersion()``` - 383 queries
  - ```.getTitle()``` - 54 queries

(This is best left for architecture meeting but I do have a relatively easy solution to fix this for *reads* not writes.  Specifically: retrieving dataset version metadata in a single query as a Map and then navigating to the needed attribute that way. Had tested it on the miniverse in Nov/Dec.)

**Comment by raprasad** [comment]
This is best left for architecture meeting but there is a relatively easy solution to fix this for *reads* not writes. (credit FB 10+ years ago)  Specifically:
  - Save 2nd version of dataset version metadata as JSON string.  
     - acts as "cached" copy--saved each time you save the dataset with published versions being static
  - Retrieve it in a single query as a Map
  - Navigate to the needed attribute
  - Had tested it on the miniverse in Nov/Dec.)

**Comment by raprasad** [comment]
Ayy!! I may split this into a second issue.

### Scenario: 
  - Go to Dataverse page
  - Click on Files facet - Test data has 13 files

#### Query count in develop: ```1,781```
#### Query count in 4.6.1: ```124```

This is not directly related to the code identified above--which works on datasets--not files.  As a sanity test, I commented out the``` DatasetUtil``` code.

**Comment by pdurbin** [comment]
> String title = dataset.getLatestVersion().getTitle();

@raprasad thanks for finding this. It's only used or debugging and should be a trivial change to replace it with the id of the dataset. This reminds me of #2110.

**Comment by pdurbin** [comment]
@raprasad @landreev @sekmiller @kcondon and I talked a fair amount about this issue today, wrote some code and did some testing.

Where there's smoke there's fire and I think there's consensus that @raprasad is on to something but in this issue about thumbnail performance. He's probably right about performance problems with the files facet in #3858 as well but I haven't personally taken a look at that. Thank you, @raprasad for caring about performance an raising a flag before we unleash 4.6.2 onto the world. We don't want to mar an otherwise fine release with performance problems.

Removing the expensive call to `dataset.getLatestVersion().getTitle()` is trivial, a one-line change that I put in pull request #3860. @kcondon did a little preliminary investigation and it seems to help. I'm confident it won't break anything. It just has to do with logging.

We decided that we'd like to take another swing at avoiding the expensive "go try to find a suitable thumbnail" logic from executing over and over. If you look at the dataset thumbnail issue at #3559 you'll see on following item on the todo list was struck:

> - ~When `thumbnailfile_id` is null, run the automatic thumbnail selection logic and persist the value of the file id to `thumbnailfile_id`. Progress in a66fce2.~ We decided not to to this after all.

We decided to take another swing at persisting the automatically-selected thumbnail in pull request #3862, which builds on the one-line fix above.

Meanwhile, there's a parallel effort to bring back the performance optimizations for automatically selecting thumbnails that existed in 4.6.1 in the `3852-query-600` branch at https://github.com/IQSS/dataverse/tree/3852-query-600 . This is a more extensive refactoring. There's no pull request as of this writing but I'm sure that code review would be appreciated.

So, to summarize, there are three branches with varying levels of change and scope:

- `3852-thumbnail-perf1` (pull request #3860) is a one-line change that fixes an obvious performance problem in the context of logging where a dataset id is just as good as a dataset title.
- `3852-thumbnail-perf2` (pull request #3862) is a relatively small change that eliminates subsequent calls to the expensive automatic thumbnail selection currently in the `develop` branch (cb0e38f) by persisting id of the file (if found!) to the database.
- [`3852-query-600`](https://github.com/IQSS/dataverse/tree/3852-query-600) is a relatively large change that brings back the faster automatic thumbnail selection that existed in 4.6.1 that I removed in the interest of making the code testable.

I hope this helps!

**Comment by landreev** [comment]
It wasn't just the "faster automatic thumbnail selection" that was removed. 
The last time we had to address the performance of the search/dataverse page, it was determined that we could not afford to instantiate any dvobjects via EJB (i.e., to look up DataFiles, Dataverses, Datasets (and versions) with ...ServiceBean.find(id). The EJB cascading was resulting in the numbers of queries that were making the performance of the page unsatisfactory. So the optimizations that I created made the page work without using such EJB instances - by making it obtain everything needed for display either from SOLR, or (for things that are not indexed) by direct native queries to the database. 
I didn't necessarily remember all the details of that project; and I did miss the fact that those optimizations were ripped out of the application in 4.6.2, and that the page went back to populating the search result objects with full EJB instances of datasets, versions, etc. Which degraded the performance of the page to where it was before the optimizations. 
Kudos to Raman for noticing it at the last moment. And I acknowledge my share of the responsibility for the fact that it wasn't discovered sooner. 
The pull request I made (https://github.com/IQSS/dataverse/pull/3864, this is "Raman's branch" - 3852-query-600) attempts to put the optimizations back. But, it's not a trivial change - arguably, a more serious change than should be made to the code on the last day before a release, lol. 

This the message I posted on slack on Friday: 

```
OK, I pushed some code into "Raman's branch" (3852-query-600). There's a serious amount of changes there: I pushed back the native queries-based optimizations that were lost in 4.6.2; but, I had to tinker with that code in order to make it support the features added in 4.6.2 - uploaded logo files, and forced "used generic thumbnail" dataset flag... In other words, there is a lot to re-test. Let's discuss this first thing on Tuesday. I'm seeing 2 choices: A) re-test all the dataset-level thumbnail-related logic carefully, merge when ready; this would mean that we assume that this regression is important enough to (potentially) delay the release; OR B) Seeing how the regression was discovered so late in the release cycle, release 4.6.2 as is (with Phil's quick fix removing that title lookup perhaps?) and fixing the performance in the next release (or in a dedicated 4.6.3?)
```

Also, similar optimizations had been made to the dataset page (the "readonly mode"). Kevin's quick tests on Friday suggested that my fixes brought the performance of the dataverse page roughly to where it was in 4.6.1; but the dataset page (that I haven't touched) was still slower. I'm assuming that the degradation of performance we are seeing there is mostly due to the code added to select that little dataset-level thumbnail for the top of the page (which, again, instantiates the version, and then instantiates all the datafiles in the version, etc.). So, ideally, that optimized code for selecting the dataset thumbnail should be moved from the searchincludefragment, and into some ViewScope wrapper, where it could be used by the dataset page as well. At the same time, the performance degradation of the dataset page should not be as severe on account of this as that on the dataverse page - because there is always only one dataset to retrieve the thumbnail for (Kevin's perliminary test results appeared to confirm that somewhat; but it should of course be tested more and more thoroughly. So, once again, let's talk about it tomorrow.

**Comment by landreev** [comment]
@kcondon 
The auto-selection of the dataset thumbnail should be consistent between the dataverse and dataset pages now. 
When the thumbnail gets auto-selected by the optimized code used by the dataverse page, it will be persisted in the db.

**Comment by landreev** [comment]
I updated the code in the branch; 
It should speed up the dataset page, among other things, as it is now using the new thumbnail service wrapper with the optimized code for looking up the dataset-level logo. 

Re: our conversation about what needs to be re-tested: 

- We've tested the 3 types of dataset-level thumbnail selections - uploaded logo file; auto-selection; assigned thumbnail; 
- persistence of auto-selected thumbnail; 
- same thumbnail should appear as the dataset logo on the dataset page, and on the search card;
- for the file-level search cards, it looks like the prudent way is to re-test separately every type of file for which we generate thumbnails

**Comment by landreev** [comment]
Also, make sure the search cards haven't lost any information, not just thumbnail-related stuff. 
Things I can think of: 
- harvesting descriptions for harvested objects;
- tabular tags and categories.

**Comment by landreev** [comment]
@kcondon a (somewhat) important thing: please note that the dataverse page should have gone back to that mode where it first fully renders *without thumbnails* (with all default icons, that is), and then initializes the thumbnails in a second pass. That was an early optimization from Gustavo, that was also lost in 4.6.2.

**Comment by landreev** [comment]
merged with develop

**Comment by landreev** [comment]
Made one more PR, per Gustavo's review (https://github.com/IQSS/dataverse/pull/3878). 
If  if a datafile is assigned as the thumbnail for a dataset, but

has been restricted since then, the assignment will be removed when the dataset gets published.
(somewhat similarly to how the map layer for a restricted file gets removed inside PublishDatasetCommand).
